File: Android/7/MovieApplication/app/src/main/java/edu/asu/msse/gnayak2/movieapplication/Movie.java
package edu.asu.msse.gnayak2.movieapplication;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;

/**
 * Copyright 2016 Gowtham Ganesh Nayak,
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * <p>
 * Purpose: This class holds all the information related to a movie.
 * <p>
 * SER598 Mobile Applications
 * see http://pooh.poly.asu.edu/Mobile
 *
 * @author Gowtham Ganesh Nayak mailto:gnayak2@asu.edu
 * @version March 2016
 */

public class Movie implements Serializable {
    private String id;
    private String title;
    private String year;
    private String rated;
    private String released;
    private String runTime;
    private String genre;
    private String actors;
    private String plot;
    private String poster;

    /**
     * Constructor
     *
     * @params title String
     * @params year String
     * @params rated String
     * @params released String
     * @params runtime String
     * @params genre String
     * @params actors String
     * @params plot String
     * @params poster String
     * @returns Object
     */
    public Movie(String title, String year, String rated, String released, String runtime, String genre, String actors, String plot, String poster) {
        this.id = UniqueIDGenerator.getUniqueId();
        this.title = title;
        this.year = year;
        this.rated = rated;
        this.released = released;
        this.runTime = runtime;
        this.genre = genre;
        this.actors = actors;
        this.plot = plot;
        this.poster = poster;
    }

    /**
     * Constructor
     *
     * @params id String
     * @params title String
     * @params year String
     * @params rated String
     * @params released String
     * @params runtime String
     * @params genre String
     * @params actors String
     * @params plot String
     * @params poster String
     * @returns Object
     */
    public Movie(String id, String title, String year, String rated, String released, String runtime, String genre, String actors, String plot, String poster) {
        this.id = id;
        this.title = title;
        this.year = year;
        this.rated = rated;
        this.released = released;
        this.runTime = runtime;
        this.genre = genre;
        this.actors = actors;
        this.plot = plot;
        this.poster = poster;
    }

    /**
     * Constructor
     *
     * @param jsonString JSON string.
     * @returns Ojbect
     */

    public Movie(String jsonString) {
        try {
            JSONObject jsonObj = new JSONObject(jsonString);
            this.title = jsonObj.getString("Title");
            this.year = jsonObj.getString("Year");
            this.rated = jsonObj.getString("Rated");
            this.released = jsonObj.getString("Released");
            this.runTime = jsonObj.getString("Runtime");
            this.actors = jsonObj.getString("Actors");
            this.plot = jsonObj.getString("Plot");
            this.genre = jsonObj.getString("Genre");
            this.poster = jsonObj.getString("Poster");
            this.id = UniqueIDGenerator.getUniqueId();
            Log.w(this.getClass().getSimpleName(), "Successfully extracrted all the information from JSON String");
        } catch (JSONException e) {
            Log.w("Json String: ", jsonString);
            Log.w(this.getClass().getSimpleName(), "Error extracting the information from JSON String");
        }
    }

    public String getTitle() {
        return this.title;
    }

    public String getYear() {
        return this.year;
    }

    public String getRated() {
        return this.rated;
    }

    public String getReleased() {
        return this.released;
    }

    public String getRunTime() {
        return this.runTime;
    }

    public String getGenre() {
        return this.genre;
    }

    public String getActors() {
        return this.actors;
    }

    public String getPlot() {
        return this.plot;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setYear(String year) {
        this.year = year;
    }

    public void setRated(String rated) {
        this.rated = rated;
    }

    public void setReleased(String released) {
        this.released = released;
    }

    public void setRunTime(String runTime) {
        this.runTime = runTime;
    }

    public void setGenre(String genre) {
        this.genre = genre;
    }

    public void setActors(String actors) {
        this.actors = actors;
    }

    public void setPlot(String plot) {
        this.plot = plot;
    }

    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setPoster(String poster) {
        this.poster = poster;
    }

    public String getPoster() {
        return this.poster;
    }

    /**
     * Converts a movie object to JSON String
     *
     * @return String
     */

    public String toJsonString() {
        String ret = "";
        try {
            JSONObject jo = new JSONObject();
            jo.put("Title", this.title);
            jo.put("Year", this.year);
            jo.put("Rated", this.rated);
            jo.put("Released", this.released);
            jo.put("Runtime", this.runTime);
            jo.put("Actors", this.actors);
            jo.put("Genre", this.genre);
            jo.put("Plot", this.plot);
            ret = jo.toString();
        } catch (Exception ex) {
            android.util.Log.w(this.getClass().getSimpleName(),
                    "error converting to/from json");
        }
        return ret;
    }
}

File: iOS/9/MovieApplication/MovieApplication/Movie.swift
/*
 * Copyright 2016 Gowtham Ganesh Nayak,
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Purpose: Assignment 9 Submission
 * This example fetches movie descriptions from the json rpc server to initialize
 * the core data and also streams the media from the media server.
 * SER 598 Mobile Applications
 * @author gnayak2@asu.edu
 * @version April 2016
 */
import Foundation

// This class is Movie Class which hold movie descriptions
class Movie {
    var id:String
    var title: String
    var year: String
    var rated: String
    var released: String
    var actors: String
    var genre: String
    var plot: String
    var runtime:String
    var media: String
    
    
    // this is from OMDB so set media to some default value
    init (jsonString: String) {
        self.title = ""
        self.rated = ""
        self.released = ""
        self.actors = "-"
        self.genre = ""
        self.plot = "-"
        self.year = ""
        self.runtime = ""
        self.media = ""
        self.id = NSUUID().UUIDString
        
        if let data: NSData = jsonString.dataUsingEncoding(NSUTF8StringEncoding) {
            do{
                let dict = try NSJSONSerialization.JSONObjectWithData(data, options:.MutableContainers) as! [String:AnyObject]
                
                let movieFound = dict["Response"] as! String
                
                if movieFound == "True" {
                    self.title = dict["Title"] as! String
                    self.year = dict["Year"] as! String
                    self.rated = dict["Rated"] as! String
                    self.released = dict["Released"] as! String
                    self.genre = dict["Genre"] as! String
                    self.actors = dict["Actors"] as! String
                    self.plot = dict["Plot"] as! String
                    self.runtime = dict["Runtime"] as! String
                    self.media = "No Media"
                } else {
                    self.title = "No movie Found"
                }
                
            } catch {
                print("unable to convert to dictionary")
                
            }
        }
    }
    
    // this is from rpc server
    init (jsonString: String, fromRPCServer:Bool) {
        self.title = ""
        self.rated = ""
        self.released = ""
        self.actors = "-"
        self.genre = ""
        self.plot = "-"
        self.year = ""
        self.runtime = ""
        self.media = "No Media"
        self.id = NSUUID().UUIDString
        
        if fromRPCServer == true {
            if let data: NSData = jsonString.dataUsingEncoding(NSUTF8StringEncoding) {
                do{
                    let dict = try NSJSONSerialization.JSONObjectWithData(data, options:.MutableContainers) as! [String:AnyObject]
                        self.title = dict["Title"] as! String
                        self.year = dict["Year"] as! String
                        self.rated = dict["Rated"] as! String
                        self.released = dict["Released"] as! String
                        self.genre = dict["Genre"] as! String
                        self.actors = dict["Actors"] as! String
                        self.plot = dict["Plot"] as! String
                        self.runtime = dict["Runtime"] as! String
                        self.media = dict["Filename"] as! String
                    
                } catch {
                    print("unable to convert to dictionary")
                    
                }
            }
        }
    }
    
    
    // In this case id is present meaning movie is present so get the media and set it
    init(id:String, title:String, year:String, rated:String, released:String, runtime:String, genre:String, actors:String, plot:String, media:String) {
        self.title = title
        self.rated = rated
        self.released = released
        self.actors = actors
        self.genre = genre
        self.plot = plot
        self.year = year
        self.runtime = runtime
        self.id = id
        self.media = media
    }
    
    func toJsonString() -> String {
        var jsonStr = "";
        let dict = ["Title": self.title, "Year": year, "Rated": self.rated, "Released": self.released, "Genre": self.genre, "Actors": self.actors,"Plot": self.plot, "Runtime": self.runtime, "Media":self.media]
        do {
            let jsonData = try NSJSONSerialization.dataWithJSONObject(dict, options: NSJSONWritingOptions(rawValue: 0))
            // here "jsonData" is the dictionary encoded in JSON data
            jsonStr = NSString(data: jsonData, encoding: NSUTF8StringEncoding)! as String
        } catch let error as NSError {
            print(error)
        }
        return jsonStr
    }
    
    // This is to manually create a movie so set to default media
    init (title: String, year: String, rated: String, released: String, runtime:String, genre: String, actors: String, plot: String) {
        self.id = NSUUID().UUIDString
        self.title = title
        self.year = year
        self.rated = rated
        self.released = released
        self.runtime = runtime
        self.genre = genre
        self.actors = actors
        self.plot = plot
        self.media = "No Media"
    }
    
    func toDict() -> [String: String]{
        var  movieDict:[String: String] = [String: String]()
        movieDict["Title"] = self.title
        movieDict["Rated"] = self.rated
        movieDict["Released"] = self.released
        movieDict["Actors"] =  self.actors
        movieDict["Genre"] = self.genre
            movieDict["Plot"] = self.plot
        movieDict["Year"] = self.year
        movieDict["Runtime"] = self.runtime
        movieDict["Id"] = self.id
        movieDict["Media"] = self.media
        return movieDict
    }
    
    func getMedia() -> String {
        return self.media
    }
    
    func setMedia(media:String) {
        self.media = media
    }
    
    func getId() -> String {
        return self.id
    }
    
    func setId(id:String) {
        self.id = id
    }
    
    func getRuntime() -> String {
        return self.runtime
    }
    
    func setRuntime(runtime:String) {
        self.runtime = runtime
    }
    
    func getGenre() -> String {
        return self.genre
    }
    
    func getTitle() -> String {
        return self.title
    }
    
    func getPlot() -> String {
        return self.plot
    }
    
    func getYear() -> String {
        return self.year
    }
    
    func getRated() -> String  {
        return self.rated
    }
    
    func getReleased() -> String  {
        return self.released
    }
    
    func getActors() -> String  {
        return self.actors
    }
    func setGenre(genre: String) {
        self.genre = genre
    }
    
    func setTitle(title: String){
        self.title = title
    }
    
    func setPlot(plot: String){
        self.plot = plot
    }
    
    func setYear(year: String){
        self.year = year
    }
    
    func setRated(rated: String){
        self.rated = rated
    }
    
    func setReleased(released: String) {
        self.released = released
    }
    
    func setActors(actors: String) {
        self.actors = actors
    }
}